### DISCUSSION TITLE
Artificial cutoff and clustering of token durations

### DISCUSSION THREAD

**Discussion by andrewreece** [deepgram/community#514] [General help] [answered]
### Which Deepgram product are you using?

Deepgram SDKs

### Details

We've observed two unexpected behaviors in Nova transcriptions:   
1. All word durations (i.e. `t_stop - t_start` for each word) are capped at exactly 500ms.  
2. Leading up to the 500ms cap, words tend to be assigned to duration values in 80ms intervals (i.e. 0.08s, 0.16s, 0.24s, etc).  (However not all words are assigned to 80ms intervals - just most of them)

These two issues appear across 800hrs of recordings we have tested on Nova 1, Nova 2-EA, and Nova 2 (GA) models.  

Attached is an audio file of a single multichannel recording that demonstrates this behavior, and Python code (DG SDK v2) to reproduce the results posted below, for Nova 2-GA.  (Strictly speaking it's not directly reproducible, as we load the audio file from a private S3 URL, but it's the same file as what is attached here.)  

It may be obvious that word duration distributions of natural spoken language do not adhere to a discretized and capped distribution, but if it's helpful, for comparison I'm also posting the distribution of word durations from the same conversation using AWS Transcribe - you can see how there is a more natural continuous curve to durations (many of which exceed 500ms).  

We are a little stumped as to what's causing this behavior - but without a fix we essentially are unable to use Deepgram STT models in scientific publications.  Any guidance much appreciated!  

Nova 1   
```
# most common token durations
0.16    1917
0.24    1828
0.50    1204
0.32    1007
0.40     531
0.08     491
0.48     304
0.25      19
0.12      14
0.38      11  
```
![image](https://github.com/deepgram/community/assets/329129/c6d0f0b7-3d83-4c25-a9ff-19954bcb4412)

Nova 2-EA  
```
# most common token durations
0.16    1843
0.24    1771
0.50    1349
0.32     947
0.08     608
0.40     517
0.48     298
0.12      30
0.30      17
0.44       9
```
![image](https://github.com/deepgram/community/assets/329129/1f65af48-bad8-4ddb-b3e7-c2a57640a615)

Nova 2-GA  
```
# most common token durations
0.16    1860
0.24    1793
0.50    1206
0.32     976
0.08     604
0.40     519
0.48     293
0.12      22
0.28       9
0.25       7
```
![image](https://github.com/deepgram/community/assets/329129/d8d54a58-43d1-4ad2-9c3e-cee241827549)

For comparison, AWS Transcribe:  
```
# most common token durations
0.01    301
0.11    204
0.19    198
0.15    198
0.13    192
0.14    188
0.17    180
0.09    173
0.18    171
0.12    170
```
![image](https://github.com/deepgram/community/assets/329129/15ea9d1d-199c-43bc-83f4-6bc631c0fdad)

### If possible, please attach your code or paste it into the text box.

```
# python 3.11, Deepgram SDK v2
import asyncio
from deepgram import Deepgram
import pandas as pd

async def get_dg_transcript(dg_client, s3_uri):
    res = await dg_client.transcription.prerecorded(
        source = {'url': s3_uri}, 
        options = {
            'model': 'nova-2',
            'smart_format': False,
            'filler_words': True,
            'punctuate': True,
            'multichannel': True,
        }
    )
    return res
    
task = asyncio.create_task(get_dg_transcript(dg_client, s3_uri), name="tx")
done, pending = await asyncio.wait({task})

if task in done:
    tx = task.result()

durations = []
for channel in tx['results']['channels']:
    for word in channel['alternatives'][0]['words']:
        durations.append(word['end'] - word['start'])

d = pd.Series(durations)
d.shape # (7372,)
d.max() # 0.5
(d == d.max()).sum() # 1201 words (16%) hit the 0.5 cutoff

# rounding for simplicity; some values are more like 0.15999 instead of 0.16, but clustering issue is the same
d.round(2).value_counts().head(10) 

# plot histogram
d.plot(kind='hist', bins=100)
```

### If possible, please attach an example audio file to reproduce the issue.

The file is too big for an attachment, so I put it on Google Drive here: https://drive.google.com/file/d/1DqXxujMPNrWTM24aLtLECjoiEgFTibNN/view?usp=sharing

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by scottstephenson** [comment, marked as answer]
Hey Andrew,  great writeup. You are right that they are discretized. We chose to do 80ms for this model. And we chose (ideal or not) to limit word lengths to 0.5s.

Of course, anything that is computed by a computer has to be discretized, and it's just a question of how much. Typical speech system discretization starts around 10ms and can go up from there. The length of a spoken word on average is around 200-300ms (which lines up with your AWS Transcribe measurement and the DG histograms). It is true that words fall on a true continuum in timing and duration with real human speech and it is true that with standard sampling rates you could hope for precision something like 1/sampling_rate. I bet you can also guess (or already know!) that the signal there would be really washed out and we'd all be agreeing that it's basically a toss up at those time resolutions with shot noise and all the other things that happen.

This doesn't explain the AWS relative smoothness, I know. They surely use a different scheme to determine the duration that works better for your high precision use case (for instance a higher time resolution VAD on top of their transcription — although I am definitely not speaking for them). I haven't done the analysis but a quick one would tell you at what level they do discretize. (I'd start with a technique similar to that used to measure the cosmic microwave background, but 1D 2 point functions, since you won't inherently be told where the boundaries they chose are.) You could try a technique like that yourself with a VAD on top of Deepgram output (we don't have a reference for it but it's not an extremely difficult thing to do on individual files).

Re: the plot from AWS; I bet you aren't hyped that AWS predicts a lot of zero word length words either (it's the highest predicted bin). But such is life, there are tradeoffs everywhere.

I understand that this might impact the scientific study of voice that you are executing and it's a bummer that our method for measuring/defining word boundaries doesn't line up with the assay you are doing. The system does a lot of things well but not what you are looking for in this case. In general, our system is set up for business use cases that don't need high precision word duration prediction, so the vast majority of the time a reasonably sub_word length duration discretization (like 80ms) is acceptable. Also, you could just change your bin size to 80ms and cut off at 500ms and they'd look similar 😝  (just kidding—I mean they _would_ look similar, but I doubt you want to do that 😄 ).

I am very interested in what it is that you are measuring and publishing about though. What is the investigation? If it's not something you want to discuss on Github, I totally understand. Feel free to contact me on Linkedin if you would like. ([my linkedin link](https://www.linkedin.com/in/scott-stephenson-/))

**Comment by andrewreece** [comment]
Hi Scott,

Thanks for the clear and thorough response - much appreciated.  Mystery solved!  I definitely appreciate how there're always tradeoffs to be made with precision vs efficiency.

As for our research at BetterUp Labs, we are currently refining our approach to conversational turn segmentation, a follow-up from our release of the [CANDOR corpus](https://www.science.org/doi/10.1126/sciadv.adf3197) earlier last year.  In reporting the characteristics of our new model, part of that entails demonstrating how our model generates distributions of key measures like turn duration and inter-speaker interval that are well-calibrated with extant literature on natural speech patterns in conversation.  That's where we run into limitations with the Nova models, particularly with the 500ms cap, as it effectively removes the upper tail of the word-length distribution.  (While 10ms units of measurement are more or less a standard in the literature, it's easier to imagine getting by with 80ms cuts than it is to have to explain why all our words stop on a dime at 500ms to reviewers.)

Working on this project has given me a much greater appreciation for the complexities of STT model design - sending big kudos to you and the whole Deepgram team for pushing the boundaries of innovation in this space.  It's quite an accomplishment!
